herd/mem.ml

```diff
@@ -991,6 +1005,21 @@ let match_reg_events es =
                 else c)
               map_load)
           stores (map_load_init loads) in
+
+      (* Add init write for system registers when one does not
```

Comment:
Can you please remind why you need to add this?

Reply:
Tests that make use of memory locations have to mention the location in the initial state (perhaps assign it to a register, since otherwise it would not be accessible by the test). As soon as that happens, the location becomes part of the "observed_locs" set. If the `-initwrites` flag is set to true, this creates a write event to that location and an `S.Store ew` is added to the reads-from map (where `ew` is the generated write event). If `-initwrites` is false, an `S.Init` is added to the reads-from map, as per the `map_load_init` function above.

The same thing happens with registers when they are part of the observed locations (mentioned in the init state or the postcondition). However, we also want to be able to use registers with default initialisation values (0) _without_ having to mention them in the init state. Since there is no way for us to know when generating the init writes that a register would be used in the code (since the event structure, with all the loads and stores of a program, is not generated at that point), a read operation of the uninitialised register would encounter no mapping in the reads-from map. This code adds an `S.Init` read-from, so that the default value can be used for uninitialised registers.